be/src/olap/memory/mem_sub_tablet.cpp

```diff
@@ -0,0 +1,247 @@
+            RETURN_IF_ERROR(_columns[cid]->create_writer(&_writers[cid]));
+        }
+    }
+    _temp_hash_entries.reserve(8);
+
+    // setup stats
+    _write_start = GetMonoTimeSecondsAsDouble();
+    _num_insert = 0;
+    _num_update = 0;
+    _num_update_cell = 0;
+    return Status::OK();
+}
+
+Status MemSubTablet::apply_partial_row_batch(PartialRowBatch* batch) {
+    while (true) {
+        bool has_row = false;
+        RETURN_IF_ERROR(batch->next_row(&has_row));
+        if (!has_row) {
+            break;
+        }
+        RETURN_IF_ERROR(apply_partial_row(*batch));
+    }
+    return Status::OK();
+}
+
+Status MemSubTablet::apply_partial_row(const PartialRowBatch& row) {
+    DCHECK_GE(row.cur_row_cell_size(), 1);
+    const ColumnSchema* dsc;
+    const void* key;
+    // get key column and find in hash index
+    // TODO: support multi-column row key
+    row.cur_row_get_cell(0, &dsc, &key);
+    ColumnWriter* keyw = _writers[1].get();
+    // find candidate rowids, and check equality
+    uint64_t hashcode = keyw->hashcode(key, 0);
+    _temp_hash_entries.clear();
+    uint32_t newslot = _write_index->find(hashcode, &_temp_hash_entries);
+    uint32_t rid = -1;
+    for (size_t i = 0; i < _temp_hash_entries.size(); i++) {
+        uint32_t test_rid = _temp_hash_entries[i];
+        if (keyw->equals(test_rid, key, 0)) {
+            rid = test_rid;
+            break;
+        }
+    }
+    // if rowkey not found, do insertion/append
+    if (rid == -1) {
+        rid = _row_size;
+        // add all columns
+        //DLOG(INFO) << StringPrintf"insert rid=%u", rid);
+        for (size_t i = 0; i < row.cur_row_cell_size(); i++) {
+            const void* data;
+            RETURN_IF_ERROR(row.cur_row_get_cell(i, &dsc, &data));
+            uint32_t cid = dsc->cid();
+            if (!_writers[cid]) {
+                RETURN_IF_ERROR(_columns[cid]->create_writer(&_writers[cid]));
+            }
+            RETURN_IF_ERROR(_writers[cid]->insert(rid, data));
+        }
+        _write_index->set(newslot, hashcode, rid);
+        _row_size++;
+        if (_write_index->need_rebuild()) {
+            scoped_refptr<HashIndex> new_index;
+            // TODO: trace memory usage
+            size_t new_capacity = _row_size * 2;
+            while (true) {
+                new_index = rebuild_hash_index(new_capacity);
+                if (new_index) {
+                    break;
+                } else {
+                    new_capacity += 1 << 16;
+                }
+            }
+            _write_index = new_index;
+        }
+        _num_insert++;
+    } else {
+        // rowkey found, do update
+        // add non-key columns
+        for (size_t i = 1; i < row.cur_row_cell_size(); i++) {
+            const void* data;
+            RETURN_IF_ERROR(row.cur_row_get_cell(i, &dsc, &data));
+            uint32_t cid = dsc->cid();
+            if (cid > _schema->num_key_columns()) {
+                if (!_writers[cid]) {
```

Comment:
we prefer `_writers[cid] == nullptr` for pointer compare

Reply:
fixed